Convert Atmel sam3x series MCU to use ASF

mnkp opened the pull request:
This patch converts Atmel sam3x MCU series to use register header files from Atmel Software Framework (ASF) library. By using ASF different Atmel SAM MCU series can use common device drivers. Deprecation of duplicate drivers will be done in subsequent patches. Minimum set of changes was performed to support ASF, additional work will be necessary to fully integrate sam3x into Atmel SAM family tree.

mnkp committed:
arch: sam3x: Use ASF library

This patch converts Atmel sam3x MCU series to use register
header files from Atmel Software Framework (ASF) library.
By using ASF different Atmel SAM MCU series can use common
device drivers.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
arch: atmel_sam: enhance gpio driver to support sam3x

This patch adds support for remaining Atmel SAM MCU series,
like sam3x to the common GPIO driver. After this update
full SAM MCU family should be supported by the driver.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

galak commented on boards/arm/arduino_due/board.h:
```diff
@@ -7,6 +7,8 @@
 #ifndef __INC_BOARD_H
 #define __INC_BOARD_H
 
+#ifndef _ASMLANGUAGE
```
this should probably be done in soc.h, not here.

(should probably also send a patch to fixup SAME70 to do the same thing

mnkp replied:
Good point, I'll make the change (also for SAME70).

tbursztyka commented on arch/arm/soc/atmel_sam/common/soc_gpio.c:
```diff
@@ -44,10 +47,22 @@ static void configure_input_attr(Pio *pio, u32_t mask, u32_t flags)
+#elif defined PIO_DIFSR_P0
+			pio->PIO_DIFSR = mask;
+#else
+#error "Unsupported Atmel SAM MCU series"
```
Can't you just do that check at the beginning of the file, once, instead? (if !defined(...) && !defined(...) #error "..." #endif )
Here it's repeating #error and makes the code quite ugly.

mnkp replied:
I'll add the following code at the beginning of the file and remove remaining #error messages.
```
#if    (!defined(PIO_IFSCER_P0) && !defined(PIO_DIFSR_P0)) \
    || (!defined(PIO_IFSCDR_P0) && !defined(PIO_SCIFSR_P0)) \ 
    || (!defined(PIO_ABCDSR_P0) && !defined(PIO_ABSR_P0)) 
#error "Unsupported Atmel SAM MCU series"
#endif
```

mnkp replied:
Hi @tbursztyka, I've provided a fix for the issue you reported. Is there anything else I should change?

fallrisk commented:
Piotr, do you have an Arduino Due? Did you compile and test this?

fallrisk commented:
Is this from the Atmel Software Framework (Atmel) or the Advanced Software Framework (Microchip). I don't know much about licenses, but has anyone confirmed the licenses? I'm attempting to download both frameworks and see what their licenses are.

mnkp commented:
I pushed one more version to remove merge conflict in boards/arm/arduino_due/arduino_due_defconfig

jiequanx committed:
net: zoap_server: Unify Makefile and add prj file

Modify the Makefile to unify with other sample apps and
add the arduino 101 configuration file.

Signed-off-by: Wu Jiequan <jiequanx.wu@intel.com>

fallrisk reviewed (commented):
I am looking into what version of ASF we have and if it came from Atmel or Microchip and looking into the licenses. If someone determines this information before me post it here.

mnkp commented:
The ASF files for SAM3X live in ext/hal/atmel/asf/sam/include/sam3x/
They were introduced in Zephyr by commit b54f2a4. The commit message states origin and the license.

jukkar committed:
net: tcp: Handle case when RST is received in any state

We must check if we receive RST in any of the TCP states.
If we do not do this, then the net_context might leak as it
would never be released in some of the states. Receiving RST
in any TCP state is not described in TCP state diagram but is
described in RFC 793 which says in chapter "Reset Processing"
that system "...aborts the connection and advises the user and
goes to the CLOSED state."

We need to also validate the received RST and accept only those
TCP reset packets that contain valid sequence number.

The validate_state_transitions() function is also changed to
accept CLOSED state transition from various other states.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon committed:
net: tcp: Handle retransmitted packets from peer.

When we receive a packet with the sequence we already seen (and
processed), the most likely cause of it is that our ACK was lost,
and peer has to retransmit that packet. Then, we should just ACK
it, because otherwise peer will retransmit it again and again,
falling into exponential backoff and hosing the entire TCP
connection.

This makes changes to send_ack(), adding a flag to force sending
an ACK regardless of its cached status, and remove inline modifier,
as the function is big and called from many places.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

jukkar committed:
net: tcp: Always set ACK bit in RST packet

Set the ACK bit in every RST packet as that seems to be expected
by Linux for example.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

fallrisk commented:
We should rename the the directory "ext/hal/atmel/asf" to "ext/hal/atmel/dfp" because the files you keep importing are from http://packs.download.atmel.com/ and are called device family packs. These aren't part of ASF. These device family packs when downloaded show an Apache 2.0 license and the individual files all have the 2 part BSD license header. @carlescufi said that the 2-part BSD license was OK for us to use.

mnkp commented:
Thanks @fallrisk for investigating the license issue!

The files were placed in asf/ directory as originally the code come from ASF. Later we got SAME70 headers with Apache 2.0 license, those were from device family packs. Also at the time it was not clear, it's in fact still not clear, if apart from header files we are going to use any source code files. Source code files are available in ASF only. I'm generally in favor of not using any ASF source code as its quality is low, but if someone needs to add quickly let's say USB support using ASF may be the best course of action. Thus, we probably shouldn't exclude this possibility.

My original suggestion when adding SAM E70 header files was not to use ASF name or asf/ folder at all but rather place all the files in ext/hal/atmel/. Vendors tend to change names of their support libraries every couple of years and we would also avoid problems like the one we're just discussing.

I'm not convinced we should rename asf/ directory just yet, but I will do as requested. In any case I propose to merge this commit as it is, assuming there are no other issues, and fix header files location later.

fallrisk commented:
Don't change it unless one of the main people agree with me. I apologize for it looking like I was wanting you to change that directory name. I want the managers to look at what I found and make the determination. Right now your PR looks good to me and I will approve it as soon as one of the managers just looks at my consideration.

mnkp commented:
Hi @fallrisk, no need to apologize. The issue you raised is certainly worth discussing. You've presented your point of view, I presented mine. At the end we all (including maintainers), hopefully come to some fruitful conclusion.

It's quite unfortunate that in open source project like Zephyr getting acquainted with a fellow coder starts with reviews. Pointing out their mistakes and differences in opinion. Nevertheless, from what I've seen so far, we are all friendly crowd.

galak commented:
We need to get Atmel to release updated SDK files with apache license.  We probably should never have allowed the import of the SAM3X files as the license isn't technically acceptable to Zephyr.

So until we resolve that (I have email requests into the Atmel / Microchip guys).  This should not be merged.

Commit:
xtensa: run simulator in turbo mode

This greatly increases the speed in which test cases are evaluated
on the Xtensa simulator.

The cycle limit parameter was removed, we don't currently need it
since sanitycheck handles killing test cases if they time out.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: fifo: extend cancel timeout limit

A non-tickless system with 10ms granularity was occasionally
taking up to 70ms for the cancellation to be propagated back.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

jukkar committed:
net: tcp: Add TCP statistics support

We did not collect any TCP statistics before but this commit
changes that.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

galak commented:
Can you rebase this and push against the 'arm' branch.

galak commented:
My previous comments about license can be ignored, didnt realize that the pack ASF files are just BSD style licensed so they are ok.

mnkp commented:
Great, I'm happy to hear the license issue was cleared!

I can push this PR against 'arm' branch, no problem. I would like to ask however if this is a one time/special request or I should push all future Atmel SAM related commits against 'arm'? Following the email "Move to Github completed and some issue" from @nashif  I was convinced we don't work with subsystem branches anymore and all push requests should be made against master. 'arm' branch is currently the only one where there is still some development going on, everyone else is pushing to master.

jukkar committed:
net: stats: Add UDP checksum stats update function

A function updating UDP checksum failure statistics was missing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: stats: Activate RPL and MLD stats if needed

If user has enabled RPL and STATISTICS, then enable RPL
statistics by default as that is probably what user want.
Same thing for MLD statistics.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: stats: IPv6 multicast listener daemon stats not printed

The MLD statistics were collected but never printed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

nashif committed:
tests: net: reduce timeout of testcase

Use a reasonable timeout for the testcase.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: remove color output from xml

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

jukkar committed:
net: context: Shorten too long line

The commit "net: tcp: Handle retransmitted packets from peer"
introduced over 80 character line that was missed in review.
Fixing it now.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

galak commented:
For the arm work I want to keep doing things on a branch, so this isn't a one time request, but going forward.  We will work to improve the process associated with branches, but will still require from time to time rebases.

mnkp commented:
I have rebased the commits on 'arm' branch and pushed to github but it looks like it does not work as expected. Github still wants to merge it into master, along some 20 or so unrelated commits. I've since rebased it back against master and re-pushed. Maybe we can merge it exceptionally into 'master'?

galak committed:
Merge branch 'arm' into sam3x

galak commented:
I was able to change it to arm

galak commented:
Close this as I pushed a new version that doesn't have the upstream master commits that I will merge.
